File: 2023/day3.py
import re

if __name__ == "__main__":
    with open('day3.data', 'r') as file:
        data = file.read().split('\n')[0:-1]

    solution_part1, solution_part2 = 0, 0

    for line_index, line in enumerate(data):
        current_number = ''
        keep_number = False
        # part1
        for char_index, char in enumerate(line):
            if '0' <= char <= '9':
                current_number += char
                min_char_index = 0 if char_index == 0 else char_index - 1
                max_char_index = len(line) - 1 if char_index == len(line) - 1 else char_index + 1
                adjacent_char_on_prev_line = (
                    line_index > 0
                    and (
                        re.match('[^0-9.]', data[line_index - 1][min_char_index])
                        or re.match('[^0-9.]', data[line_index - 1][char_index])
                        or re.match('[^0-9.]', data[line_index - 1][max_char_index])
                    )
                )
                adjacent_char_on_current_line = (
                    re.match('[^0-9.]', data[line_index][min_char_index])
                    or re.match('[^0-9.]', data[line_index][char_index])
                    or re.match('[^0-9.]', data[line_index][max_char_index])
                )
                adjacent_char_on_next_line = (
                    line_index < len(data) - 1
                    and (
                        re.match('[^0-9.]', data[line_index + 1][min_char_index])
                        or re.match('[^0-9.]', data[line_index + 1][char_index])
                        or re.match('[^0-9.]', data[line_index + 1][max_char_index])
                    )
                )
                if adjacent_char_on_prev_line or adjacent_char_on_current_line or adjacent_char_on_next_line:
                    keep_number = True
            if not '0' <= char <= '9' or char_index == len(line) - 1:
                if keep_number and not current_number == '':
                    solution_part1 += int(current_number)
                    keep_number = False
                current_number = ''
        # part2
        asterisks = re.finditer('\*', line)
        for asterisk in asterisks:
            adjacent_numbers = []
            position = asterisk.start()
            numbers_on_prev_line = [] if line_index == 0 else re.finditer('[0-9]+', data[line_index - 1])
            numbers_on_current_line = re.finditer('[0-9]+', line)
            numbers_on_next_line = [] if line_index == len(data) - 1 else re.finditer('[0-9]+', data[line_index + 1])
            for number in numbers_on_prev_line:
                if abs(number.start() - position) <= 1 or abs(number.end() - position - 1) <= 1:
                    adjacent_numbers.append(int(data[line_index - 1][number.start():number.end()]))
            for number in numbers_on_current_line:
                if number.start() == position + 1 or number.end() == position:
                    adjacent_numbers.append(int(line[number.start():number.end()]))
            for number in numbers_on_next_line:
                if abs(number.start() - position) <= 1 or abs(number.end() - position - 1) <= 1:
                    adjacent_numbers.append(int(data[line_index + 1][number.start():number.end()]))
            if len(adjacent_numbers) == 2:
                solution_part2 += adjacent_numbers[0] * adjacent_numbers[1]

    print('Solution Part 1: %d' % solution_part1)
    print('Solution Part 2: %d' % solution_part2)

File: 2023-rust/src/solutions/day3.rs
use crate::utils::load_data;
use regex::Regex;


///!
///! # WARNING
///! This part of the code was originally written in Python and was just moved over to Rust.
///! Code quality might not be optimal
///!


pub fn run(test: bool) -> (i64, i64) {
    let data = load_data(3, test);
    let data = data.split('\n').collect::<Vec<_>>();
    let mut part1 = 0;
    let mut part2 = 0;
    for (line_index, line) in data.iter().enumerate() {
        if line.is_empty() { continue; }
        let mut current_number = String::new();
        let mut keep_number = false;

        // PART 1
        for (char_index, char) in line.chars().enumerate() {
            if char.is_numeric() {
                current_number.push(char);
                let min_char_index = if char_index == 0 { 0 } else { char_index - 1 };
                let max_char_index = if char_index == line.len() - 1 {
                    line.len() - 1
                } else {
                    char_index + 1
                };
                let special_char = Regex::new("[^0-9.]").unwrap();
                let adjacent_char_on_prev_line = line_index > 0 && (
                    special_char.is_match(&data[line_index - 1]
                        .chars()
                        .nth(min_char_index)
                        .unwrap()
                        .to_string()
                    ) || special_char.is_match(&data[line_index - 1]
                        .chars()
                        .nth(char_index)
                        .unwrap()
                        .to_string()
                    ) || special_char.is_match(&data[line_index - 1]
                        .chars()
                        .nth(max_char_index)
                        .unwrap()
                        .to_string()
                    )
                );
                let adjacent_char_on_current_line = special_char.is_match(&data[line_index]
                    .chars()
                    .nth(min_char_index)
                    .unwrap()
                    .to_string()
                ) || special_char.is_match(&data[line_index]
                    .chars()
                    .nth(char_index)
                    .unwrap()
                    .to_string()
                ) || special_char.is_match(&data[line_index]
                    .chars()
                    .nth(max_char_index)
                    .unwrap()
                    .to_string()
                );
                let adjacent_char_on_next_line = line_index < data.len() - 1 && (
                    special_char.is_match(&data[line_index + 1]
                        .chars()
                        .nth(min_char_index)
                        .unwrap()
                        .to_string()
                    ) || special_char.is_match(&data[line_index + 1]
                        .chars()
                        .nth(char_index)
                        .unwrap()
                        .to_string()
                    ) || special_char.is_match(&data[line_index + 1]
                        .chars()
                        .nth(max_char_index)
                        .unwrap()
                        .to_string()
                    )
                );
                if adjacent_char_on_prev_line || adjacent_char_on_current_line || adjacent_char_on_next_line {
                    keep_number = true;
                }
            }
            if !char.is_numeric() || char_index == line.len() - 1 {
                if keep_number && !current_number.is_empty() {
                    part1 += current_number.parse::<i64>().unwrap();
                    keep_number = false;
                }
                current_number.clear();
            }
        }

        // PART 2
        let asterisks = Regex::new("\\*").unwrap();
        let numeric = Regex::new("[0-9]+").unwrap();
        let asterisks = asterisks.find_iter(line);
        for asterisk in asterisks {
            let position = asterisk.start();
            let numbers_on_prev_line = if line_index == 0 {
                vec![]
            } else {
                numeric.find_iter(data[line_index - 1])
                    .filter(|m| (m.start() as i64 - position as i64).abs() <= 1 || (m.end() as i64 - position as i64 - 1).abs() <= 1)
                    .map(|m| data[line_index - 1][m.start()..m.end()].parse::<i64>().unwrap())
                    .collect::<Vec<_>>()
            };
            let numbers_on_current_line = numeric.find_iter(line)
                .filter(|m| m.start() == position + 1 || m.end() == position)
                .map(|m| line[m.start()..m.end()].parse::<i64>().unwrap())
                .collect::<Vec<_>>();
            let numbers_on_next_line = if line_index == data.len() - 1 {
                vec![]
            } else {
                numeric.find_iter(data[line_index + 1])
                    .filter(|m| (m.start() as i64 - position as i64).abs() <= 1 ||(m.end() as i64 - position as i64 - 1).abs() <= 1)
                    .map(|m| data[line_index + 1][m.start()..m.end()].parse::<i64>().unwrap())
                    .collect::<Vec<_>>()
            };
            if numbers_on_prev_line.len() + numbers_on_current_line.len() + numbers_on_next_line.len() == 2 {
                part2 += numbers_on_prev_line.iter().product::<i64>() * numbers_on_current_line.iter().product::<i64>() * numbers_on_next_line.iter().product::<i64>();
            }
        }
    }

    return (part1 as i64, part2 as i64);
}